nexus/types/src/quiesce.rs

```diff
@@ -1110,4 +1274,143 @@ mod test {
+        // Act like we assigned some sagas.
+        reassignment.reassignment_done(
+            SagaReassignmentDone::ReassignedAllAsOf(blueprint2_id, true),
+        );
+        // We're not fully drained because we haven't recovered those sagas.
+        assert!(!qq.is_fully_drained());
+        // So the fully drained blueprint is the one from before.
+        assert_eq!(qq.fully_drained_blueprint(), Some(blueprint1_id));
+
+        // Start a recovery pass.  Pretend like we found a saga.
+        // We'll use a oneshot channel to emulate the saga completion future.
+        let (tx, rx) = tokio::sync::oneshot::channel();
+        let recovery = qq.recovery_start();
+        let pending = recovery.record_saga_recovery(*SAGA_ID, &SAGA_NAME);
+        let consumer_completion = pending.saga_completion_future(
+            async { rx.await.expect("cannot drop this before dropping tx") }
+                .boxed(),
+        );
+        recovery.recovery_done(true);
+
+        // We're still not fully drained because we haven't finished that saga.
+        assert!(!qq.is_fully_drained());
+        // So the fully drained blueprint is the one from before.
+        assert_eq!(qq.fully_drained_blueprint(), Some(blueprint1_id));
+
+        // Fully drained case 2: saga completion causes us to become fully
+        // drained.
+        //
+        // Complete the saga.
+        tx.send(saga_result()).unwrap();
+        let _ = consumer_completion.await;
+        // Now, we should be fully drained up to the new blueprint.
+        assert!(qq.is_fully_drained());
+        assert_eq!(qq.fully_drained_blueprint(), Some(blueprint2_id));
+
+        // Fully drained case 3: saga recovery causes us to become fully
+        // drained.
+        //
+        // For this case, imagine that we think we may have re-assigned
+        // ourselves some sagas, but recovery completes with no sagas
+        // outstanding.
+        let blueprint3_id = BlueprintUuid::new_v4();
+        let reassignment = qq.reassignment_start();
+        assert_eq!(qq.fully_drained_blueprint(), Some(blueprint2_id));
+        // Act like we assigned some sagas.
+        reassignment.reassignment_done(
+            SagaReassignmentDone::ReassignedAllAsOf(blueprint3_id, true),
+        );
+        assert!(!qq.is_fully_drained());
+        assert_eq!(qq.fully_drained_blueprint(), Some(blueprint2_id));
+
+        // Quick check: failed recovery changes nothing.
+        qq.recovery_start().recovery_done(false);
+        assert!(!qq.is_fully_drained());
+        assert_eq!(qq.fully_drained_blueprint(), Some(blueprint2_id));
+
+        // Successful recovery with no sagas running means we're fully drained
+        // as of the new blueprint.
+        qq.recovery_start().recovery_done(true);
+        assert_eq!(qq.fully_drained_blueprint(), Some(blueprint3_id));
+        assert!(qq.is_fully_drained());
+
+        // Fully drained case 3: quiescing itself causes us to immediately
```

Comment:
```suggestion
        // Fully drained case 4: quiescing itself causes us to immediately
```

Could some of these cases become different, smaller tests?